### ISSUE TITLE
port CGI.escapeHTML optimization

### ISSUE THREAD

**Issue by ahorek** [jruby/jruby#5937]
one of the recent ruby 2.7 optimization is https://github.com/ruby/ruby/pull/2226

jruby is a bit behind
```
ruby 2.7 dev
      CGI.escapeHTML      2.614k (± 9.3%) i/s -     13.108k in   5.065551s

ruby 2.5.3
      CGI.escapeHTML    761.177  (± 6.7%) i/s -      3.825k in   5.050563s

jruby 9.2.9.0 dev
      CGI.escapeHTML    224.765  (± 6.7%) i/s -      1.122k in   5.013482s
```

the previous java port is at https://github.com/jruby/jruby/blob/4e8bb2666a7257f0f5986800f96bb88efdd6acbd/core/src/main/java/org/jruby/ext/cgi/escape/CGIEscape.java#L64

porting the new algorithm should improve numbers

the optimization is especially important for rails, erb engine does use it for html escaping:
https://api.rubyonrails.org/classes/ERB/Util.html#method-c-h

**Comment by headius** [comment]
As long as there's no functionality change, I think we can pull this into 9.2.9 (still 2.6 compat, etc).

**Comment by headius** [comment]
Oops, for some reason I thought this was a PR. We'll merge it whenever someone does it.

**Comment by ahorek** [comment]
sry, no PR yet, but I can work on it.

just one question, the optimized java version doesn't seem to be used now
https://github.com/jruby/jruby/blob/4e8bb2666a7257f0f5986800f96bb88efdd6acbd/core/src/main/java/org/jruby/ext/cgi/escape/CGIEscape.java#L352

was it removed intentionally in https://github.com/jruby/jruby/issues/4531 ?

**Comment by headius** [comment]
There's still issues with supering up a complicated include+prepend+extend hierarchy, so we'll aim to get that and this wrapped up in 9.2.10.

**Comment by enebo** [comment]
@ahorek https://github.com/jruby/jruby/pull/6777  

Can you try this port on this branch?  I believe it will work now.  If so we can land this for 9.3 instead of 9.4.

**Comment by enebo** [comment]
Fixed prepend issue in PR #6777 and impld in #6778
